### ISSUE TITLE
Investigate the best design to improve performance of zerocracy

### ISSUE THREAD

**Issue by victornoel** [zerocracy/farm#2150]
Following the investigation in #2119, here are some ideas about improving things.

They should be studied to find the better design to implement in the medium/long term.

In the medium term:
* we should consider differentiating read only operation from write operations, it is not without reasons that `ReadWriteLock` have a specific behaviour related to that. This would prevent all the web access from blocking stuffs too much. You could even prioritize write operation then.

In the longer term:
*  we don't really want a lock, we want a queue: there are tasks to do but they can be done only one by one per project. So you need one queue per project. It is possible to implement queues in the database (to answer the same requirements that motivated implementing the lock in the database) and it will be much more efficient and easy to understand. There exists some resources on that on the internet, some are related to advisory locks (see also #2149 also).
* I would even ask (but this is a huge endeavour, I would say a rewrite): isn't it the time to move to real even-oriented architecture that handles all this complexity for you instead of (badly, because we aren't experts) reinventing the wheel?

**Comment by 0crat** [comment]
@g4s8[/z](https://www.0crat.com/u/g4s8) please, pay attention to this issue

**Comment by g4s8** [comment]
@rultor release, tag=`0.49.16`

**Comment by 0pdd** [comment]
@victornoel the puzzle [#2165](https://github.com/zerocracy/farm/issues/2165) is still not solved.

**Comment by g4s8** [comment]
@victornoel thanks for the ticket. I started new implementation of async logic - each project has its own thread now, then we will check that resource can be accessed only by project thread and remove synchronizations after #2156

**Comment by 0crat** [comment]
Job `gh:zerocracy/farm#2150` is not assigned, can't get performer

**Comment by 0crat** [comment]
This job is not in scope

**Comment by g4s8** [comment]
@rultor release, tag=`0.49.18`

**Comment by 0pdd** [comment]
@victornoel the puzzle [#2170](https://github.com/zerocracy/farm/issues/2170) is still not solved; solved: [#2165](https://github.com/zerocracy/farm/issues/2165).

**Comment by 0pdd** [comment]
@victornoel 2 puzzles [#2170](https://github.com/zerocracy/farm/issues/2170), [#2171](https://github.com/zerocracy/farm/issues/2171) are still not solved; solved: [#2165](https://github.com/zerocracy/farm/issues/2165).
